Barlt logfile extractor

noahmacaskill opened the pull request:
Created a log file metadata extractor for Frontier Labs. Just pulled all of the information I saw there (all SD card related). Some fields may be more relevant than others (definitely the case on our end).

Included a --support-dirs option with this, basically giving the user the option to specify where to search for any support files (on our end they're often lumped together in the same directory). If it isn't used the default is to search in the directory of the audio file (and all subdirectories).

atruskie commented:
@noahmacaskill I did start reviewing this but I had to evacuate due to the floods that hit Brisbane. I'll get back to this ins few days 

If you're stuck for work you can try to work on the bar lt header extraction from the FLAC comment block in the files. You can branch off of this branch if you want to build on your work

atruskie commented on src/MetadataUtility/Metadata/FrontierLabs/LogFileExtractor.cs:
```diff
@@ -0,0 +1,64 @@
+
+            while (i < lines.Length)
+            {
+                if (lines[i].Contains("SD Card :"))
+                {
+                    break;
+                }
```
BAR LTs support up to 4 SD cards

atruskie replied:
This also means you won't know which card belongs to which recording unless you correlate them some how?

noahmacaskill replied:
It seems like new log files are created for each SD card, but couldn't find anything in the Frontier Labs documentation to confirm this. Instead I collect every instance of an SD card in a log file. If they aren't all representing the same SD Card, then I correlate them if possible (only works with newer firmware, where the filename is printed each time a new recording is started).

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/FLLogFile.cs:
```diff
@@ -0,0 +1,61 @@
+    using System.Text.RegularExpressions;
+    using LanguageExt;
+
+    public static class FLLogFile
```
We prefer to avoid initialisms. Maybe just add this to a `FrontierLabs` namespace & folder like we do for other file types

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/FLLogFile.cs:
```diff
@@ -0,0 +1,61 @@
+
+            foreach (string logFile in logFiles)
+            {
+                string[] lines = information.FileSystem.File.ReadAllLines(logFile);
```
Reading all lines in a file to check a header seems very expensive

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/FLLogFile.cs:
```diff
@@ -0,0 +1,61 @@
+            {
+                string[] lines = information.FileSystem.File.ReadAllLines(logFile);
+
+                if (lines[2].Equals(FrontierLabsLogString))
```
I would scan the first few lines for the header - all sorts of weird behaviour could cause this header to be off by a line

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/FLLogFile.cs:
```diff
@@ -0,0 +1,61 @@
+                    {
+                        if (line.Contains(information.FileSystem.Path.GetFileName(information.Path)))
+                        {
+                            information.KnownSupportFiles.Add("Log file", logFile);
```
Make the dictionary key a constant

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/FLLogFile.cs:
```diff
@@ -0,0 +1,61 @@
+                        }
+                    }
+
+                    // Check if the log file is the only log file in its directory (and all subdirectories), and if the audio file is in its directory (or a subdirectory)
+                    List<string> files = information.FileSystem.Directory.GetFiles(information.FileSystem.Path.GetDirectoryName(logFile), "*", SearchOption.AllDirectories).ToList();
+
+                    if (files.Count(s => Regex.IsMatch(s, ".*logfile.*txt")) == 1 && files.Count(s => s.EndsWith(information.FileSystem.Path.GetFileName(information.Path))) == 1)
+                    {
+                        information.KnownSupportFiles.Add("Log file", logFile);
+                        return true;
+                    }
```
I assume the reason for this complexity is because the support dirs could find multiple log files?

Let's try and simplify this a bit.

atruskie reviewed (commented):
So I was a little surprised you did this first. I figured grabbing metadata from the header comment in the files would be more important? Doesn't matter though.

I also hadn't fully thought out the design of the support files. Be prepared for some rearchitecting.

I think FL don't ever put their log files more than a few levels above the source file. So, I'd suggest our implementation should scan parent directories for the target log files (with a maximum height of say three parent directories). This would omit the need for the support directories command line argument.

I see that the main reason you wrote this extractor is to get the SD card information. That information is in the header comment as well. We should write adaptors for both in case either is missing.

noahmacaskill commented:
Did a pretty big overhaul of the structure for this one. The biggest change being, any time a log file has to be read in, we extract every piece of useful information we can from it, then cache it in a static [list](https://github.com/QutEcoacoustics/emu/blob/ec331a08b37daedb083b9d7e7c184ffeb6bef160/src/MetadataUtility/Metadata/TargetInformation.cs#L77).

atruskie commented on src/MetadataUtility/Metadata/FrontierLabs/LogFileExtractor.cs:
```diff
@@ -0,0 +1,98 @@
+    {
+        public const int MetadataOffset = 39;
+        public const int SerialNumberLineOffset = 6;
+        public const string LogFileKey = "Log file";
```
```suggestion
        public const string LogFileKey = "Frontier Labs Log File";
```

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/SupportFile.cs:
```diff
@@ -0,0 +1,57 @@
+
+        public static List<string> FindSupportFiles(TargetInformation information, string pattern)
+        {
+            List<string> logFiles = new List<string>();
```
```suggestion
            List<string> supportFiles = new List<string>();
```

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/SupportFile.cs:
```diff
@@ -0,0 +1,57 @@
+        public static List<string> FindSupportFiles(TargetInformation information, string pattern)
+        {
+            List<string> logFiles = new List<string>();
+            string fileDirectory = information.FileSystem.Path.GetDirectoryName(information.Path);
+            string searchDirectory = information.FileSystem.Path.GetDirectoryName(information.Path);
```
```suggestion
            string fileDirectory = information.FileSystem.Path.GetDirectoryName(information.Path);
            string searchDirectory = fileDirectory;
```

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/SupportFile.cs:
```diff
@@ -0,0 +1,57 @@
+
+            int i = 0;
+
+            while (i++ < 3 && (logFiles = information.FileSystem.Directory.GetFiles(searchDirectory, pattern, SearchOption.AllDirectories).ToList()).Length() == 0)
```
```suggestion
            const maxHeight = 3;
            while (i++ < maxHeight  && !(logFiles = information.FileSystem.Directory.GetFiles(searchDirectory, pattern, SearchOption.AllDirectories).Any())
```

atruskie replied:
Personally, I'd just put the complexities of this into the body. You can always break or continue inside the loop body

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/SupportFile.cs:
```diff
@@ -0,0 +1,57 @@
+                }
+            }
+
+            if (logFiles.Length() == 1)
```
```suggestion
            var first = logFiles.FirstOrDefault();
            if (first is not null)
```

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/SupportFile.cs:
```diff
@@ -0,0 +1,57 @@
+        }
+
+        /// <summary>
+        /// Extracts all useful information out of the log file.
```
```suggestion
        /// Extracts all useful information out of the support file.
```

atruskie commented on src/MetadataUtility/Metadata/SupportFiles/SupportFile.cs:
```diff
@@ -0,0 +1,57 @@
+{
+    public abstract class SupportFile
+    {
+        public string File { get; set; }
```
is this the file contents or the file path?

noahmacaskill replied:
File path, updated the name here.

atruskie commented on src/MetadataUtility/Metadata/TargetInformation.cs:
```diff
@@ -64,10 +65,16 @@ public Stream FileStream
+        /// Gets list of each identified support file
+        /// Support files are log files, configuration files, etc.
+        /// </summary>
+        public static List<SupportFile> KnownSupportFiles { get; } = new List<SupportFile>();
```
ok I'd suggest reverting this back to a dictionary keyed off of the support file type.

I assume you switched this to a list because your search may return multiple log files per target? The hope would be that is not the case and we can discriminate the correct log file per target.

However, if for some reason there are multiple support files of the same type for a single target, then other solutions might be more useful (like bundling the collection into the `SupportFile` class).

atruskie replied:
Actually, scratch my last; why do we have two collections at all? What's the difference between KnownSupportFiles and TargetSupportFiles?

It looks like one is a list and one is a dictionary? If that is the sole difference you can collapse these into the one type and use a dictionary like a list in most cases.

atruskie replied:
Oh, it's static. I see. You mentioned that. Forgive my iterative drama.

Static, in short, is a no-no. If there is any kind of parallelism or async activity that shares this data structure you're going to run into race conditions. I'd suggest doing a two-phase scan. I'll add more comments where I see this used.

atruskie commented on src/MetadataUtility/Models/MemoryCard.cs:
```diff
@@ -0,0 +1,39 @@
+    /// </summary>
+    public record MemoryCard
+    {
+        public string SDFormatType { get; init; }
+
+        public uint SDManufacturerID { get; init; }
+
+        public string SDOEMID { get; init; }
+
+        public string SDProductName { get; init; }
+
+        public float SDProductRevision { get; init; }
+
+        public uint SDSerialNumber { get; init; }
+
+        public string SDManufactureDate { get; init; }
+
+        public uint SDSpeed { get; init; }
+
+        public uint SDCapacity { get; init; }
+
+        public uint SDWrCurrentVmin { get; init; }
+
+        public uint SDWrCurrentVmax { get; init; }
+
+        public uint SDWriteB1Size { get; init; }
+
+        public uint SDEraseB1Size { get; init; }
+    }
```
Great putting this into its own structure. Can we remove all the `SD` prefixes?

atruskie commented on test/MetadataUtility.Tests/Metadata/FrontierLabs/LogFileExtractorTests.cs:
```diff
@@ -0,0 +1,74 @@
+        [ClassData(typeof(FixtureHelper.FixtureData))]
+        public async void CanProcessFilesWorks(FixtureModel model)
+        {
+            var result = await this.subject.CanProcessAsync(model.ToTargetInformation(this.RealFileSystem));
+
+            // we can process any file that has Frontier Lab log files
+            var expected = model.Process.Contains("FrontierLabsLogFileExtractor");
+            Assert.Equal(expected, result);
```
I think this test is pretty flexible and powerful. I like this. Well done us.

atruskie commented on test/MetadataUtility.Tests/Metadata/FrontierLabs/LogFileExtractorTests.cs:
```diff
@@ -0,0 +1,74 @@
+                    ti,
+                    this.Recording);
+
+                recording.MemoryCard.SDFormatType.Should().Be(model.SDFormatType);
+                recording.MemoryCard.SDManufacturerID.Should().Be(model.SDManufacturerID);
+                recording.MemoryCard.SDOEMID.Should().Be(model.SDOEMID);
+                recording.MemoryCard.SDProductName.Should().Be(model.SDProductName);
+                recording.MemoryCard.SDProductRevision.Should().Be(model.SDProductRevision);
+                recording.MemoryCard.SDSerialNumber.Should().Be(model.SDSerialNumber);
+                recording.MemoryCard.SDManufactureDate.Should().Be(model.SDManufactureDate);
+                recording.MemoryCard.SDSpeed.Should().Be(model.SDSpeed);
+                recording.MemoryCard.SDCapacity.Should().Be(model.SDCapacity);
+                recording.MemoryCard.SDWrCurrentVmin.Should().Be(model.SDWrCurrentVmin);
+                recording.MemoryCard.SDWrCurrentVmax.Should().Be(model.SDWrCurrentVmax);
+                recording.MemoryCard.SDWriteB1Size.Should().Be(model.SDWriteB1Size);
+                recording.MemoryCard.SDEraseB1Size.Should().Be(model.SDEraseB1Size);
+            }
```
https://fluentassertions.com/objectgraphs/

```suggestion
                recording.MemoryCard.Should().BeEquivalentTo(memoryCard);
```

(we may have to construct a MemoryCard variable first)

atruskie reviewed (changes requested):
Just a note: be aware that every `.ToList()` that isn't needed can add a fair bit of a performance hit because a new array is created and the contents are copied in each time. 
Try to use the IEnumerable type and its methods instead. 

OK, this is a great improvement.

I think there's some comments to clean up. I like the `SupportFile` abstraction. 

What I'd like to see now is the support files scanned for an initialized in the `Metadata` command when the `TargetInformation`s are being created. That's the right time to share state between each one (i.e. caching known support files and assigning them to multiple target informations) since the concurrency there is limited.

Like last time, update our metadata docs to indicate which fields you're adding support for in this PR ( we might add partial support since this information is found both in the support file and in the file header).

atruskie commented:
BTW, I made other changes to metadata.md you'll need to merge in: https://github.com/QutEcoacoustics/emu/commit/c28c0893e0d6c71546457f2631c977e329596da3

noahmacaskill commented:
Moved the scanning of support files into Metadata (when we create each TargetInformation). Kept the static list of known support files in target information, but it's not used during any of the asynchronous parts.

When searching for support files, I had been searching not just parent directories of each target, but all subdirectories as well. I reverted that here, mostly because I think it overcomplicates things, and isn't necessary. Thoughts?

In searching all subdirectories though I did find that the [normal FLAC file](https://github.com/QutEcoacoustics/emu/blob/e7df090cb3f8f80da0484e583290fa009c9cef87/test/Fixtures/FL_BAR_LT/3.14_Normal/20191026T000000+1000_REC.flac) shares the same log file as the [glitchy signal log file](https://github.com/QutEcoacoustics/emu/blob/e7df090cb3f8f80da0484e583290fa009c9cef87/test/Fixtures/FL_BAR_LT/3.14_GlitchySignal/logfile.txt) so I copied that log file over to the same directory as the normal file.

atruskie reviewed (changes requested):
> When searching for support files, I had been searching not just parent directories of each target, but all subdirectories as well. I reverted that here, mostly because I think it overcomplicates things, and isn't necessary. Thoughts?

At this stage I can't think of any sensor the stores support files in a directory deeper than a target audio file. So yeah, an ascending (current level included) search seems appropriate. 

> In searching all subdirectories though I did find that the [normal FLAC file](https://github.com/QutEcoacoustics/emu/blob/e7df090cb3f8f80da0484e583290fa009c9cef87/test/Fixtures/FL_BAR_LT/3.14_Normal/20191026T000000+1000_REC.flac) shares the same log file as the [glitchy signal log file](https://github.com/QutEcoacoustics/emu/blob/e7df090cb3f8f80da0484e583290fa009c9cef87/test/Fixtures/FL_BAR_LT/3.14_GlitchySignal/logfile.txt) so I copied that log file over to the same directory as the normal file.

Fair enough. I'd probably symlink it in the future, but good spot nonetheless. 


OK, we're getting close. I was really hoping to push this through today but a little bit more cleaning will set us up well for the future.

I'm somewhat obsessive about avoiding expensive operations; fast tools are good to use and more importantly scale up to very large numbers of files. Doing lots of nested loops (like scanning the same directories for the same files for every target files (that's a O^3 operation!)) is the sort of thing that would cripple performance... hence why I'm focusing on it.

noahmacaskill commented on src/MetadataUtility/Metadata/SupportFiles/FrontierLabs/LogFile.cs:
```diff
@@ -0,0 +1,165 @@
+                            TargetInformation.KnownSupportFiles.Add(logFile);
+                        }
+
+                        // If more than one log file is found, we must check for the file name in the log file
+                        // Won't apply to later firmware versions
```
We do, if an SD card gets filled up, when moving to the next SD slot it will create a new log file. We have a sensor that produced "logfile.txt", "logfile (2).txt" and "logfile (3).txt". Will add this case to the fixtures file.

noahmacaskill replied:
Actually seems this is probably just due to logfiles being collapsed into one directory on our end at some point. The sensor will produce one log file per SD slot, but will by default separate these into different directories.

Regardless, that was why I'm checking if there is one log file (where we can assume it matches the target due to directory structure) or multiple (where I then search each logfile for the specific recording name). I'll leave it up to you whether this is worth keeping or not since most people won't have this issue so long as they keep the original directory structure. If not I can keep it on a separate local copy for our purposes.

atruskie replied:
Okay, this use case seems fair enough. 

But so far we don't have any data where this is demonstrated, so the code written to select the right file will never be tested in our unit tests.

Add another fixture: two files from one sensor, one from each sd card, and two log files (that have the window copy name deduplication suffix in them)..

noahmacaskill replied:
Our WAV files are pretty huge (2GB each). I uploaded empty WAV files with the same names, since that's all we need to match with the log files. That sound okay?

noahmacaskill commented:
I also realized I was extracting the firmware version this whole time but haven't been saving it anywhere. Set it in Sensor.cs, and while I was at it extracted a couple more things on the Sensor (serial number and power source).

atruskie commented on src/MetadataUtility/Models/Sensor.cs:
```diff
@@ -20,12 +18,12 @@ public record Sensor
         /// Gets the firmware version of this sensor.
         /// </summary>
-        public string Firmware { get; init; }
+        public float Firmware { get; init; }
```
```suggestion
        public string Firmware { get; init; }
```

You cannot guarantee a  firmware version is parseable as a float. FL do adhere to the 'version is a number' convention, but most versions is the world can have multiple dotted segments.

atruskie reviewed (approved):
I didn't add sufficient instruction for adding new samples to the fixtures folder so this isn't your fault, but I do need to see license and provenance information for the files you added. I added some instructions into @JordanMercerECCC's branch you can refer to: https://github.com/QutEcoacoustics/emu/blob/f7028afe614150cc2b4e4426053be45b56f06ec2/test/Fixtures/README.md

Your method of including some small files for the log file extractor tests is sound. Good job. 
We have limited support for adding larger files (through git-lfs) but I still don't want the repository to become huge. 

Two small changes; you can make them out-of-band, or if you beat me to merging you can do them in band

atruskie commented:
Actually I'll do the firmware version and merge. But I do need you to provide that provenance information
